Denoise profile updates

rawfiner opened the pull request:
Current state:
1- Fix anscombe transform so that the white balance coefficients are correctly taken into account
2- Adds an option in case user doesn't want a white balance adaptive transform
3- Fix non local means patch norm computation (as a result, increasing patch size does not require to increase the force anymore -> easier to use)
4- Add a slider to control the weight of central pixel in non local means patch comparison metric

Basically I currently use non local means like this :
- I put the patch size to 4
- I increase the coarse grain noise slider until no very coarse chroma noise remains (some fine grain chrominance noise can sometimes remain)
- if image is too much blurred, I reduce the force *slightly*
- then I increase the details slider to my taste
- I fix the remaining chroma noise with the equalizer
Note that with these changes, I am mostly using only one instance of denoiseprofile, without any particular blending mode.

It works very well for medium and high iso, but is not perfect for very high iso. In such case, it can be combined with denoise bilateral to get nice results.

This is work in progress, I did not update the presets yet, and I have ongoing work to make the anscombe transform perform better.

rawfiner commented:
I updated the presets, rebased on master and force-pushed.
I will make another PR later for the anscombe changes, they take more time and may be bigger that I first thought.

TurboGit commented:
So you mean that this is ready for review/integration?

rawfiner commented:
Yes, this is ready for review

TurboGit commented:
@rawfiner : I don't like the backward compatibility GUI section. I understand that you need to keep that with a values FALSE for both boolean for previous edits. What I would do:

* solution 1:
- for old edits display the backward compatibility section with both values not selected
- this make it possible for users to tick both options (fix ascomb, fix patch)
- when both options are selected do not display this section at all, I don't think someone would want to uncheck the fix (at least we certainly don't want to encourage this).

* solution 2:
The other solution is to never show those options. Both are FALSE for old edit and we set them to TRUE automatically for new edit or when the module is reset.

If we go with solution 1, I would even have a single tick:
 [ ] fix ascom transform and patch normalisation
For old edit andf again do not display that if it has already been selected by users.

TurboGit commented on src/iop/denoiseprofile.c:
```diff
@@ -278,6 +308,41 @@ int legacy_params(dt_iop_module_t *self, const void *const old_params, const int
+    }
+    v7->scattering = v6.scattering;
+    v7->central_pixel_weight = 0.0;
+    v7->fix_anscombe = FALSE;     // don't fix anscombe to ensure backward compatibility
```
A single boolean here:
`  v7->fixed_algo`

?

TurboGit commented on src/iop/denoiseprofile.c:
```diff
@@ -1508,12 +1685,47 @@ static int process_nlmeans_cl(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop
+  // The 0.045 was derived from the old formula, to keep the
+  // norm identical when P=1, as the norm for P=1 seemed
+  // to work quite well: 0.045 = 0.015 * (2 * P + 1) with P=1.
+  float norm = .045f / ((2 * P + 1) * (2 * P + 1));
```
```
float norm;
if(d->fixed_algo)
   norm = .045f / ((2 * P + 1) * (2 * P + 1));
else
   norm = .015f / (2 * P + 1);
```

TurboGit commented on src/iop/denoiseprofile.c:
```diff
@@ -1508,12 +1685,47 @@ static int process_nlmeans_cl(struct dt_iop_module_t *self, dt_dev_pixelpipe_iop
+  // we init wb by the mean of the coeffs, which corresponds to the mean
+  // amplification that is done in addition to the "ISO" related amplification
+  float wb[4] = { wb_mean, wb_mean, wb_mean, 0.0f };
+  if(d->fix_anscombe)
```
```
if(d->fixed_algo)
{
}
else
{
}
```

TurboGit commented on src/iop/denoiseprofile.c:
```diff
@@ -1792,9 +2007,36 @@ static int process_wavelets_cl(struct dt_iop_module_t *self, dt_dev_pixelpipe_io
+  // we init wb by the mean of the coeffs, which corresponds to the mean
+  // amplification that is done in addition to the "ISO" related amplification
+  float wb[4] = { wb_mean, wb_mean, wb_mean, 0.0f };
+  if(d->fix_anscombe)
```
```
if(d->fixed_algo)
...
```

TurboGit commented on src/iop/denoiseprofile.c:
```diff
@@ -2774,24 +3086,70 @@ void gui_init(dt_iop_module_t *self)
   gtk_box_pack_start(GTK_BOX(self->widget), g->stack, TRUE, TRUE, 0);
+  gtk_box_pack_start(GTK_BOX(self->widget), g->strength, TRUE, TRUE, 0);
+
+  // add collapsable section for those extra options that are generally not to be used
```
Replace all this by a single boolean option and only displayed if (d->fixed_alfo is FALSE)

rawfiner commented:
Thank you for reviewing and testing.
I don't mind removing the backward compatibility section, it will simplify the GUI.
I do not know which solution is the best for the user:
- if we hide completely the options the user may find it confusing that from one photo (which had already a denoise profile instance) to another (a new one) he gets very different results with similar settings.
- if we show the option only in case of an old edit, the GUI code is a little bit more complicated, but the user can see something changes. Yet, it may be confusing too to see an option on an image and don't see it on another one?
As such, I do not know which solution to choose. Do you have a preference?

TurboGit commented:
I will go for 2 and if the option is named:

`[  ] migrate to new algorithm`

And in the tooltip message you can have a longer explanation about the ascomb and patch fix and the fact that when selected it won't be possible to go back later.

It will not be confusing to not see this in new edits I would say.

rawfiner commented:
Ok, thanks for the idea, I will do that :-)

rawfiner commented:
@TurboGit , should be better now.
Yet, I have some GUI issues, I don't manage yet to make the option not appear on new instances

TurboGit commented on src/iop/denoiseprofile.c:
```diff
@@ -2116,11 +2352,14 @@ void reload_defaults(dt_iop_module_t *module)
+    ((dt_iop_denoiseprofile_params_t *)module->default_params)->central_pixel_weight = 0.0f;
     ((dt_iop_denoiseprofile_params_t *)module->default_params)->strength = 1.0f;
     ((dt_iop_denoiseprofile_params_t *)module->default_params)->mode = MODE_NLMEANS;
+    ((dt_iop_denoiseprofile_params_t *)module->default_params)->fix_anscombe_and_nlmeans_norm = TRUE;
```
By default it should be FALSE, no?

TurboGit commented on src/iop/denoiseprofile.c:
```diff
@@ -2148,6 +2387,8 @@ void init(dt_iop_module_t *module)
       tmp.y[ch][k] = 0.5f;
     }
   }
+  tmp.fix_anscombe_and_nlmeans_norm = TRUE;
```
FALSE also here?

rawfiner commented:
> In mode_callback you also need to hide the (fix anscomb...) after:
> gtk_stack_set_visible_child_name(GTK_STACK(g->stack), "wavelets");

I added this in the fix_anscombe_and_nlmeans_norm_callback, I think it makes more sense than in the mode callback.
Basically, it makes the button disapear when the user click on it.

And your commit 1589f0d has solved my issue with new instances, many thanks! :-) \o/

TurboGit commented:
> I added this in the fix_anscombe_and_nlmeans_norm_callback, I think it makes more sense than in the mode callback.

I don't think this is correct. A user may want to activate and deactivate it multiple time at different zoom level to asses the changes and if he want to keep the previous algorithm or not.

rawfiner commented:
> > I added this in the fix_anscombe_and_nlmeans_norm_callback, I think it makes more sense than in the mode callback.
> 
> I don't think this is correct. A user may want to activate and deactivate it multiple time at different zoom level to asses the changes and if he want to keep the previous algorithm or not.

Ok, I have removed the commit. Do you think we should do it when the user change mode however?

TurboGit commented:
> Do you think we should do it when the user change mode however?

When changing mode we should probably keep the current status of the fix boolean.

rawfiner commented:
> > Do you think we should do it when the user change mode however?
> 
> When changing mode we should probably keep the current status of the fix boolean.

Sorry, I am not sure I understand correctly.
If I understand correctly, you mean that changing mode we should not do anything, and keep the checkbox shown if it is shown?

TurboGit commented:
Exactly. And not shown if not shown.

rawfiner commented:
> Exactly. And not shown if not shown.

Ok :-)
I was confused with the remark about the mode_callback.
Thanks for clarifying
Last pushed version should work as expected.

TurboGit commented:
Ok, UI fixed now, thanks. But the presets are not working on my side now.

rawfiner commented:
Good remark. I will update them again

rawfiner commented:
The presets should be fixed
I also updated one of the tooltip text, which was inaccurate

edgardoh commented:
This is amazing! I have never been able to get good results with this denoise until now, thanks soooo much!

TurboGit commented:
Really nice indeed! Great work rawfiner. Thanks.

rawfiner commented:
@edgardoh @TurboGit Thanks for the feedback :-)
I am glad you like it !

And thanks for the code reviews !

rawfiner commented:
@TurboGit just to check, the modifications seem to appear twice in master history (all my commits + a "merge" commit that contains all the changes) is that normal?

TurboGit commented:
I don't see that on my side.

edgardoh commented:
I'm having an issue, when exporting downscaled is like the module is disabled, there's no effect at all, if I export in HQ then things go back to normal. Is this expected?

rawfiner commented:
@edgardoh This issue was already here in 2.6 (and before).
The module still has some effect, but the effects are different than the ones we could expect: for instance there is undersmoothing.
The problem is that it is hard to adapt the algorithms to downsampling.
If we note I the original image, D the downsampling, f the denoising function, p its parameters, it is hard to find parameters p' such as:
D(f(I,p)) = f(D(I),p')
For now, the parameters p' which are estimated are not good enough.
If you want the result to be accurate with what you see at 100% zoom, you have to select the option "do high quality resampling during export", so that downsampling is down at the end of the pipe and not at the beginning.

edgardoh commented:
OK, thanks for the explanation.


El vie., 15 mar. 2019 a las 6:04, rawfiner (<notifications@github.com>)
escribió:

> @edgardoh <https://github.com/edgardoh> This issue was already here in
> 2.6 (and before).
> The module still has some effect, but the effects are different than the
> ones we could expect: for instance there is undersmoothing.
> The problem is that it is hard to adapt the algorithms to downsampling.
> If we note I the original image, D the downsampling, f the denoising
> function, p its parameters, it is hard to find parameters p' such as:
> D(f(I,p)) = f(D(I),p')
> For now, the parameters p' which are estimated are not good enough.
> If you want the result to be accurate with what you see at 100% zoom, you
> have to select the option "do high quality resampling during export", so
> that downsampling is down at the end of the pipe and not at the beginning.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/2127#issuecomment-473209526>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APq1pRisscLjF-Ju5xtkRrlJC1hq-1W6ks5vW2ItgaJpZM4bN-XY>
> .
>

edgardoh commented:
@rawfiner , I don't know if you are working on making this scale-independent, but since right now results are not usable on a downscaled image, it is possible to just skip the process? This is a heavy module, so this will make the pipe faster. Maybe choose a scale and anything below just copy input -> output.

rawfiner commented:
@edgardoh Yes, at least I will try to make it more scale-independent. I am not in favor of skipping completely the process, I would rather have a better preview. We will see if this is feasible (I have some ideas, but not tested them yet). If if does not work correctly, maybe should we tell the user to zoom at 100%, and do copies at other zoom levels, as you suggest. But I hope the preview will improve enough so that we don't need to do that.

edgardoh commented:
That sounds great, thanks!

El dom., 24 mar. 2019 a las 12:36, rawfiner (<notifications@github.com>)
escribió:

> @edgardoh <https://github.com/edgardoh> Yes, at least I will try to make
> it more scale-independent. I am not in favor of skipping completely the
> process, I would rather have a better preview. We will see if this is
> feasible (I have some ideas, but not tested them yet). If if does not work
> correctly, maybe should we tell the user to zoom at 100%, and do copies at
> other zoom levels, as you suggest. But I hope the preview will improve
> enough so that we don't need to do that.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/darktable-org/darktable/pull/2127#issuecomment-475970410>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/APq1paG_Rcf5JKYaO9iqg21bJYqWUQehks5vZ5uMgaJpZM4bN-XY>
> .
>
